src/main/scala/reporting/Converter.scala

```diff
@@ -0,0 +1,531 @@
+    } else {
+      ConstantEntry(value.toString)
+    }
+  }
+}
+
+case class RefEntry(
+    name: String,
+    fields: Map[String, (ExtractedModelEntry, Option[Rational])]
+) extends ExtractedModelEntry {
+  override def toString: String = {
+    val buf = fields
+      .map(x =>
+        s"\t${x._1}(perm: ${x._2._2.map(_.toString).getOrElse("?")}) <- ${x._2._1.toString.split("\n").mkString("\n\t")}\n"
+      )
+      .mkString("")
+    s"Ref ($name) {\n$buf}"
+  }
+  def asValueEntry = ConstantEntry(name)
+}
+
+case class NullRefEntry(name: String) extends ExtractedModelEntry {
+  override def toString = s"Null($name)"
+  def asValueEntry = ConstantEntry(name)
+}
+
+case class RecursiveRefEntry(name: String) extends ExtractedModelEntry {
+  override def toString = s"recursive reference to $name"
+  def asValueEntry = ConstantEntry(name)
+}
+
+case class VarEntry(name: String, sort: Sort) extends ExtractedModelEntry {
+  override def toString: String = name
+  def asValueEntry = ConstantEntry(name)
+}
+
+case class OtherEntry(value: String, problem: String = "")
+    extends ExtractedModelEntry {
+  override def toString = s"$value [$problem]"
+  def asValueEntry = ConstantEntry(value)
+}
+
+case class SeqEntry(name: String, values: List[ExtractedModelEntry])
+    extends ExtractedModelEntry {
+  override def toString = s"($name): [${values.map(_.toString).mkString(", ")}]"
+  def asValueEntry = ConstantEntry(name)
+}
+
+case class UnprocessedModelEntry(entry: ValueEntry)
+    extends ExtractedModelEntry {
+  override def toString = s"$entry"
+  def asValueEntry = entry
+}
+
+// processed Heap representation:
+sealed trait HeapEntry{
+  def toString: String
+}
+
+case class ExtractedHeap(entries: List[HeapEntry])
+
+case class PredHeapEntry(
+    name: String,
+    args: Seq[ExtractedModelEntry]
+) extends HeapEntry {
+  override def toString = s"$name(${args.mkString(", ")})"
+}
+case class FieldHeapEntry(
+    recv: VarEntry,
+    field: String,
+    perm: Option[Rational],
+    sort: Sort,
+    entry: ExtractedModelEntry
+) extends HeapEntry {
+  override def toString = s"$recv.$field"
+}
+
+case class UnresolvedHeapEntry(chunk: Chunk, reason: String) extends HeapEntry {
+  override def toString = s"$chunk (not further processed because: $reason)"
+}
+
+object Converter {
+
+  def getFunctionValue(
+      model: Model,
+      fname: String,
+      args: Seq[ValueEntry],
+      toSort: Sort
+  ): ExtractedModelEntry = {
+    val entry: Option[ModelEntry] = model.entries.get(fname)
+    entry match {
+      case Some(MapEntry(m, els)) =>
+        getConstantEntry(toSort, m.getOrElse(args, els))
+      case Some(m) => getConstantEntry(toSort, m)
+      case None    => OtherEntry("${fname}", "function not found in model")
+    }
+  }
+
+  def translateSort(s: Sort): String = {
+    s match {
+      case sorts.Set(els: Sort) => s"Set<${translateSort(els)}>"
+      case sorts.Ref            => "$Ref"
```

Comment:
It would be more stable w.r.t to maintenance if you didn't use a string literal here (and in several other places), which may get out of sync with the rest of Silicon. Not sure how practical it is, but you could potentially use [`TermToSMTLib2Converter`](https://github.com/viperproject/silicon/blob/master/src/main/scala/decider/TermToSMTLib2Converter.scala#L36) to go from `sorts.Ref` to `"$Ref"`.